File: src/quests/stats.py
"""Aggregate quest activity + periodic template streaks for /api/stats."""

from __future__ import annotations

from datetime import date, datetime, time, timedelta, timezone
from typing import Any
from zoneinfo import ZoneInfo

from sqlmodel import col, select
from sqlmodel.ext.asyncio.session import AsyncSession

from quests.models import Quest, QuestChangeLog, QuestStatus, QuestTemplate
from quests.periodic import resolve_tz
from quests.timeutil import ensure_utc, remaining_seconds

UTC = timezone.utc

ISSUED_KINDS = frozenset({"quest_created", "quest_appeared"})
COMPLETED_KIND = "quest_completed"
FAILED_KIND = "quest_failed"


def local_day(dt: datetime, tz: ZoneInfo) -> date:
    aware = ensure_utc(dt)
    assert aware is not None
    return aware.astimezone(tz).date()


def resolve_range(
    *,
    days: int | None,
    date_from: date | None,
    date_to: date | None,
    tz: ZoneInfo,
    now: datetime | None = None,
) -> tuple[date, date]:
    now = ensure_utc(now or datetime.now(UTC))
    assert now is not None
    today = now.astimezone(tz).date()
    end = date_to or today
    if date_from is not None:
        start = date_from
    else:
        n = max(1, int(days or 30))
        start = end - timedelta(days=n - 1)
    if start > end:
        start, end = end, start
    return start, end


def range_bounds_utc(start: date, end: date, tz: ZoneInfo) -> tuple[datetime, datetime]:
    """Inclusive local dates → half-open UTC [start, end_exclusive)."""
    start_local = datetime.combine(start, time.min, tzinfo=tz)
    end_exclusive_local = datetime.combine(end + timedelta(days=1), time.min, tzinfo=tz)
    return start_local.astimezone(UTC), end_exclusive_local.astimezone(UTC)


def instance_outcome(quest: Quest, *, now: datetime | None = None) -> str:
    """completed | miss | open — only for periods that have an instance."""
    now = ensure_utc(now or datetime.now(UTC))
    status = quest.status
    if status == QuestStatus.completed:
        return "completed"
    if status in (QuestStatus.failed, QuestStatus.delayed, QuestStatus.archived):
        return "miss"
    rem = remaining_seconds(quest.deadline_at, now=now)
    if rem is not None and rem <= 0:
        return "miss"
    return "open"


def compute_streaks(outcomes_chrono: list[str]) -> tuple[int, int]:
    """Empty periods are absent from the list → they do not break the streak."""
    longest = 0
    run = 0
    for outcome in outcomes_chrono:
        if outcome == "completed":
            run += 1
            longest = max(longest, run)
        else:
            # miss or historical open → break longest run
            run = 0

    i = len(outcomes_chrono) - 1
    while i >= 0 and outcomes_chrono[i] == "open":
        i -= 1
    current = 0
    while i >= 0 and outcomes_chrono[i] == "completed":
        current += 1
        i -= 1
    return current, longest


def empty_daily(start: date, end: date) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    d = start
    while d <= end:
        out.append({"date": d.isoformat(), "issued": 0, "completed": 0, "failed": 0})
        d += timedelta(days=1)
    return out


async def build_stats(
    session: AsyncSession,
    *,
    days: int | None = 30,
    date_from: date | None = None,
    date_to: date | None = None,
    template_id: int | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    tz = resolve_tz(None)
    now = ensure_utc(now or datetime.now(UTC))
    assert now is not None
    start, end = resolve_range(days=days, date_from=date_from, date_to=date_to, tz=tz, now=now)
    lo, hi = range_bounds_utc(start, end, tz)

    daily_map = {row["date"]: row for row in empty_daily(start, end)}

    log_result = await session.exec(
        select(QuestChangeLog).where(
            col(QuestChangeLog.at) >= lo.replace(tzinfo=None),
            col(QuestChangeLog.at) < hi.replace(tzinfo=None),
            col(QuestChangeLog.kind).in_(
                list(ISSUED_KINDS | {COMPLETED_KIND, FAILED_KIND})
            ),
        )
    )
    # Distinct quest_id per kind per local day
    seen: set[tuple[str, str, int | None]] = set()
    for row in log_result.all():
        day = local_day(row.at, tz).isoformat()
        if day not in daily_map:
            continue
        kind = row.kind
        key = (day, kind, row.quest_id)
        if key in seen:
            continue
        seen.add(key)
        bucket = daily_map[day]
        if kind in ISSUED_KINDS:
            bucket["issued"] += 1
        elif kind == COMPLETED_KIND:
            bucket["completed"] += 1
        elif kind == FAILED_KIND:
            bucket["failed"] += 1

    tmpl_result = await session.exec(
        select(QuestTemplate).order_by(col(QuestTemplate.id))
    )
    templates = [
        {"id": int(t.id), "title": t.title, "enabled": bool(t.enabled)}
        for t in tmpl_result.all()
        if t.id is not None
    ]

    chosen_id = template_id
    if chosen_id is None and templates:
        enabled = [t for t in templates if t["enabled"]]
        chosen_id = (enabled or templates)[0]["id"]

    template_block: dict[str, Any] | None = None
    if chosen_id is not None:
        template_block = await _template_stats(
            session,
            template_id=int(chosen_id),
            start=start,
            end=end,
            now=now,
        )

    return {
        "range": {"from": start.isoformat(), "to": end.isoformat()},
        "daily": [daily_map[d.isoformat()] for d in _iter_days(start, end)],
        "templates": templates,
        "template": template_block,
    }


def _iter_days(start: date, end: date):
    d = start
    while d <= end:
        yield d
        d += timedelta(days=1)


async def _template_stats(
    session: AsyncSession,
    *,
    template_id: int,
    start: date,
    end: date,
    now: datetime,
) -> dict[str, Any] | None:
    tmpl = await session.get(QuestTemplate, template_id)
    if tmpl is None or tmpl.id is None:
        return None

    result = await session.exec(
        select(Quest)
        .where(Quest.template_id == template_id)
        .where(col(Quest.period_key).is_not(None))
        .order_by(col(Quest.period_key))
    )
    instances = list(result.all())

    outcomes_all: list[str] = []
    for q in instances:
        outcomes_all.append(instance_outcome(q, now=now))
    current_streak, longest_streak = compute_streaks(outcomes_all)

    bars: list[dict[str, Any]] = []
    closed = 0
    total = 0
    for q in instances:
        key = q.period_key or ""
        if not key:
            continue
        try:
            pk = date.fromisoformat(key)
        except ValueError:
            continue
        if pk < start or pk > end:
            continue
        outcome = instance_outcome(q, now=now)
        status = q.status.value if hasattr(q.status, "value") else str(q.status)
        bars.append(
            {
                "period_key": key,
                "status": status,
                "outcome": outcome,
                "quest_id": q.id,
            }
        )
        total += 1
        if outcome == "completed":
            closed += 1

    close_rate = (closed / total) if total else 0.0
    return {
        "id": int(tmpl.id),
        "title": tmpl.title,
        "current_streak": current_streak,
        "longest_streak": longest_streak,
        "closed": closed,
        "total": total,
        "close_rate": round(close_rate, 4),
        "bars": bars,
    }

File: go/internal/store/stats.go
package store

import (
	"context"
	"database/sql"
	"os"
	"time"

	"github.com/valirum/quests/go/internal/timeutil"
)

func (s *Store) QuestIDForStep(ctx context.Context, stepID int64) (int64, error) {
	var qid sql.NullInt64
	err := s.DB.QueryRowContext(ctx, `SELECT quest_id FROM queststep WHERE id=?`, stepID).Scan(&qid)
	if err == sql.ErrNoRows || !qid.Valid {
		return 0, ErrNotFound
	}
	if err != nil {
		return 0, err
	}
	return qid.Int64, nil
}

func (s *Store) ListChangeLog(ctx context.Context, limit int, questID *int64, beforeID *int64) ([]map[string]any, error) {
	if limit < 1 {
		limit = 100
	}
	if limit > 500 {
		limit = 500
	}
	q := `SELECT id, at, kind, quest_id, title, detail, significance, revision
		FROM questchangelog WHERE 1=1`
	args := []any{}
	if questID != nil {
		q += ` AND quest_id=?`
		args = append(args, *questID)
	}
	if beforeID != nil {
		q += ` AND id < ?`
		args = append(args, *beforeID)
	}
	q += ` ORDER BY at DESC, id DESC LIMIT ?`
	args = append(args, limit)
	rows, err := s.DB.QueryContext(ctx, q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]map[string]any, 0)
	for rows.Next() {
		var id int64
		var at, kind, title, detail string
		var nqid, nrev sql.NullInt64
		var nsig sql.NullString
		if err := rows.Scan(&id, &at, &kind, &nqid, &title, &detail, &nsig, &nrev); err != nil {
			return nil, err
		}
		t, _ := timeutil.ParseFlexible(at)
		m := map[string]any{
			"id": id, "at": derefISO(&t), "kind": kind, "quest_id": nil,
			"title": title, "detail": detail, "significance": nil, "revision": nil,
		}
		if nqid.Valid {
			m["quest_id"] = nqid.Int64
		}
		if nsig.Valid {
			m["significance"] = nsig.String
		}
		if nrev.Valid {
			m["revision"] = int(nrev.Int64)
		}
		out = append(out, m)
	}
	return out, rows.Err()
}

func (s *Store) BuildStats(ctx context.Context, days int, templateID *int64) (map[string]any, error) {
	if days < 1 {
		days = 30
	}
	if days > 366 {
		days = 366
	}
	tzName := os.Getenv("QUESTS_TZ")
	if tzName == "" {
		tzName = "Europe/Moscow"
	}
	loc, err := time.LoadLocation(tzName)
	if err != nil {
		loc = time.UTC
	}
	now := timeutil.NowUTC()
	today := now.In(loc)
	end := time.Date(today.Year(), today.Month(), today.Day(), 0, 0, 0, 0, loc)
	start := end.AddDate(0, 0, -(days - 1))

	daily := []map[string]any{}
	dailyMap := map[string]map[string]any{}
	for d := start; !d.After(end); d = d.AddDate(0, 0, 1) {
		key := d.Format("2006-01-02")
		row := map[string]any{"date": key, "issued": 0, "completed": 0, "failed": 0}
		daily = append(daily, row)
		dailyMap[key] = row
	}

	lo := start.UTC()
	hi := end.AddDate(0, 0, 1).UTC()
	rows, err := s.DB.QueryContext(ctx, `
		SELECT at, kind, quest_id FROM questchangelog
		WHERE at >= ? AND at < ? AND kind IN ('quest_created','quest_appeared','quest_completed','quest_failed')`,
		timeutil.ToDBUTC(lo), timeutil.ToDBUTC(hi))
	if err != nil {
		return nil, err
	}
	seen := map[string]struct{}{}
	for rows.Next() {
		var at, kind string
		var qid sql.NullInt64
		if err := rows.Scan(&at, &kind, &qid); err != nil {
			rows.Close()
			return nil, err
		}
		t, _ := timeutil.ParseFlexible(at)
		day := t.In(loc).Format("2006-01-02")
		bucket, ok := dailyMap[day]
		if !ok {
			continue
		}
		qkey := day + "|" + kind + "|"
		if qid.Valid {
			qkey += itoa64(qid.Int64)
		}
		if _, ok := seen[qkey]; ok {
			continue
		}
		seen[qkey] = struct{}{}
		switch kind {
		case "quest_created", "quest_appeared":
			bucket["issued"] = bucket["issued"].(int) + 1
		case "quest_completed":
			bucket["completed"] = bucket["completed"].(int) + 1
		case "quest_failed":
			bucket["failed"] = bucket["failed"].(int) + 1
		}
	}
	rows.Close()

	trows, err := s.DB.QueryContext(ctx, `SELECT id, title, enabled FROM questtemplate ORDER BY id`)
	if err != nil {
		return nil, err
	}
	templates := []map[string]any{}
	var firstEnabled, firstAny *int64
	for trows.Next() {
		var id int64
		var title string
		var en int
		_ = trows.Scan(&id, &title, &en)
		templates = append(templates, map[string]any{"id": id, "title": title, "enabled": en != 0})
		if firstAny == nil {
			v := id
			firstAny = &v
		}
		if en != 0 && firstEnabled == nil {
			v := id
			firstEnabled = &v
		}
	}
	trows.Close()

	chosen := templateID
	if chosen == nil {
		if firstEnabled != nil {
			chosen = firstEnabled
		} else {
			chosen = firstAny
		}
	}
	var tmplBlock map[string]any
	if chosen != nil {
		tmplBlock, err = s.templateStats(ctx, *chosen, now)
		if err == ErrNotFound {
			return nil, err
		}
	}

	return map[string]any{
		"range":     map[string]string{"from": start.Format("2006-01-02"), "to": end.Format("2006-01-02")},
		"daily":     daily,
		"templates": templates,
		"template":  tmplBlock,
	}, nil
}

func (s *Store) templateStats(ctx context.Context, templateID int64, now time.Time) (map[string]any, error) {
	var title string
	err := s.DB.QueryRowContext(ctx, `SELECT title FROM questtemplate WHERE id=?`, templateID).Scan(&title)
	if err != nil {
		return nil, ErrNotFound
	}
	rows, err := s.DB.QueryContext(ctx, `
		SELECT id, status, period_key, deadline_at FROM quest
		WHERE template_id=? AND period_key IS NOT NULL ORDER BY period_key`, templateID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	outcomes := []string{}
	bars := []map[string]any{}
	closed, total := 0, 0
	for rows.Next() {
		var id int64
		var status, pk string
		var deadline sql.NullString
		_ = rows.Scan(&id, &status, &pk, &deadline)
		outcome := "open"
		switch status {
		case "completed":
			outcome = "completed"
		case "failed", "delayed", "archived":
			outcome = "miss"
		default:
			if deadline.Valid {
				t, e := timeutil.ParseFlexible(deadline.String)
				if e == nil && !t.After(now) {
					outcome = "miss"
				}
			}
		}
		outcomes = append(outcomes, outcome)
		bars = append(bars, map[string]any{
			"period_key": pk, "status": status, "outcome": outcome, "quest_id": id,
		})
		total++
		if outcome != "open" {
			closed++
		}
	}
	cur, longest := computeStreaks(outcomes)
	rate := 0.0
	nClosed, nComp := 0, 0
	for _, o := range outcomes {
		if o == "open" {
			continue
		}
		nClosed++
		if o == "completed" {
			nComp++
		}
	}
	if nClosed > 0 {
		rate = float64(nComp) / float64(nClosed)
	}
	return map[string]any{
		"id": templateID, "title": title,
		"current_streak": cur, "longest_streak": longest,
		"closed": closed, "total": total, "close_rate": rate,
		"bars": bars,
	}, nil
}

func computeStreaks(outcomes []string) (current, longest int) {
	run := 0
	for _, o := range outcomes {
		if o == "completed" {
			run++
			if run > longest {
				longest = run
			}
		} else {
			run = 0
		}
	}
	i := len(outcomes) - 1
	for i >= 0 && outcomes[i] == "open" {
		i--
	}
	for i >= 0 && outcomes[i] == "completed" {
		current++
		i--
	}
	return current, longest
}
